wemake_python_styleguide/visitors/ast/builtins.py

```diff
@@ -125,6 +124,128 @@ def _check_modulo_patterns(
             self.add_violation(consistency.ModuloStringFormatViolation(node))
 
 
+@final
+class WrongFormatStringVisitor(base.BaseNodeVisitor):
+    """Restricts usage of ``f`` strings."""
+
+    _valid_format_index: ClassVar[AnyNodes] = (
+        ast.Str,
+        ast.Num,
+        ast.Name,
+    )
+    _dual_components: ClassVar[AnyNodes] = (
+        ast.Call,
+        ast.Subscript,
+    )
+    _chainable_types: ClassVar[AnyNodes] = (
+        ast.Call,
+        ast.Subscript,
+        ast.Attribute,
+    )
+
+    def visit_JoinedStr(self, node: ast.JoinedStr) -> None:
+        """
+        Forbids use of ``f`` strings and too complex ``f`` strings.
+
+        Raises:
+            FormattedStringViolation
+            TooComplexFormattedStringViolation
+
+        """
+        self.add_violation(consistency.FormattedStringViolation(node))
+        self.generic_visit(node)
+        self._check_complex_formatted_string(node)
+
+    def _check_complex_formatted_string(self, node: ast.JoinedStr) -> None:
+        # Whitelists all simple uses of f strings
+        # Checks if list, dict, function call with no parameters or variable
+        has_formatted_components = any(
+            isinstance(comp, ast.FormattedValue) for comp in node.values
+        )
+        if not has_formatted_components:
+            # If no formatted values
+            self.add_violation(
+                complexity.TooComplexFormattedStringViolation(node),
+            )
+            return
+
+        for string_component in node.values:
+            if isinstance(string_component, ast.FormattedValue):
+                # Test if possible chaining is invalid
+                format_value = string_component.value
+                if self._is_valid_formatted_value(format_value):
+                    continue
+                # Everything else is too complex
+                self.add_violation(
+                    complexity.TooComplexFormattedStringViolation(node),
+                )
+                break
+
+    def _is_valid_formatted_value(self, format_value: ast.AST) -> bool:
+        if isinstance(format_value, (ast.Call, ast.Subscript, ast.Attribute)):
```

Comment:
Can be replaced with `self._chainable_types`

Reply:
Yes but then it won't pass Mypy since `format_value` will still be considered AST, and `_is_valid_chaining(self, format_value: AnyChainable)` takes AnyChainable as parameter.

Following error will be displayed:
```
wemake_python_styleguide/visitors/ast/builtins.py:186: error: Argument 1 to "_is_valid_chaining" of "WrongFormatStringVisitor" has incompatible type "AST"; expected "Union[Attribute, Subscript, Call]"
```

Reply:
This will work if you move `_any_chainable` to module level. And declare it `final`. 
That's a bug in mypy, but we can live with it. Just drop a comment on why it is done this way.